### ISSUE TITLE
Anomaly "Uncaught exception Invalid_argument("Lexer: location function")." with custom entries

### ISSUE THREAD

**Issue by maximedenes** [coq/coq#9358]
The following script:

```coq
Declare Custom Entry expr.
Notation "[ e ]" := e (e custom expr at level 2).
Check [ fun x => x ].
```

gives the anomaly in the title.

Tested in master 2e231b0eb3d107aedc7a26dc173d6ac00cb3473f

**Comment by herbelin** [comment]
This is a tricky bug, at the border of camlp5 (gramlib) and Coq. 

The problem is a bug in printing the error message. The intended error is `[constr:expr level 2] expected after '[' (in [constr:operconstr])` but the computation of the location of the error fails.

It is assumed that, when such an error arrives, the next token has been consumed. This invariant is however wrong in the case of a token followed by an entry with only empty levels as it is the case in the bug report.

`Plexing` says: `The type of a function giving the location of a token in the source from the token number in the stream (starting from zero).` Here the function is `CLexer.loct_func` which recovers a loc formerly stored in a hash-table at the time the next token is peeked.

Several possible solutions, such as:
- We ensure in camlp5/gramlib that token number n of a stream has actually been peeked before asking for the location of the n-th token (easiest fix, a reasonable one).
- We bypass this in `CLexer` by enforcing a peek ourself before computing the loc.
- We get rid of the side effect keeping the loc attached to the token rather than storing it in an hash-table. Would it be costly? (Anyway, If there is an efficiency problem, there are other sources of improvement which we could think of anyway, e.g. looking if a specific `IDENT "foo"` match in log-time rather than linear time, for `"foo"` a specific constant.)

Side remark: `Print Grammar expr` should say something and does not. There is something to check or fix also here.

**Comment by ejgallego** [comment]
Thanks for the detailed analysis @herbelin , 3 looks like the most sensible thing to do [we could profit from `CAst.t`].

I don't think that would be too costly IIUC.

**Comment by Janno** [comment]
Here's another piece of code that triggers this error message:
```coq
Declare Custom Entry bla.
Notation "[ t ]" := (t) (at level 0, t custom bla at level 0).
Notation "t ]" := (t) (in custom bla at level 0, t constr).
Check [.
```
Is this the same underlying failure?

The same error message is printed for 
```coq
Check [ 1 ] ].
```

**Comment by herbelin** [comment]
Fixed in #9604 (but following the shortest path, option 1).
